[DX-1024, TT-11563] Updated mock response (#4074)
TykTechnologies/tyk-docs: tyk-docs/content/product-stack/tyk-gateway/middleware/mock-response-tyk-classic.md

---
title: Using the Mock Response middleware with Tyk Classic APIs
date: 2024-01-31
description: "Using the Mock Response middleware with Tyk Classic APIs"
tags: ["mock response", "middleware", "per-endpoint", "Tyk Classic", "Tyk Classic API"]
---

The [Mock Response]({{< ref "product-stack/tyk-gateway/middleware/mock-response-middleware" >}}) middleware allows you to configure Tyk to return a response for an API endpoint without requiring an upstream service. This can be useful when creating a new API or making a development API available to an external team.

When working with Tyk Classic APIs, this middleware is executed at the start of the request processing chain. Thus an endpoint with the mock response middleware will not be authenticated, will not process other middleware configured for the API (neither API nor endpoint level) and will have no analytics created.  It will simply return the configured response for any request made to that endpoint.

> The middleware is configured in the Tyk Classic API Definition, either manually within the `.json` file or from the API Designer in the Tyk Dashboard.

Author: ```suggestion
The middleware is configured in the Tyk Classic API Definition. You can do this via the Tyk Dashboard API or in the API Designer.
```

                        }
                    }          
                }
            }
        ]
    }
}
```

In this example the mock response middleware has been configured for HTTP `GET` requests to the `/anything` endpoint. The [allow list]({{< ref "advanced-configuration/transform-traffic/endpoint-designer#allowlist" >}}) middleware has been enabled for this endpoint and is case sensitive, so calls to `GET /Anything` will not return the mock response.

A call to `GET /anything` would return:

```
HTTP/1.1 200 OK
X-Example-Header: foobar
Date: Wed, 31 Jan 2024 16:21:05 GMT
Content-Length: 30
Content-Type: text/plain; charset=utf-8

This is the mock response body
```

## Configuring the middleware in the API Designer
You can use the API Designer in the Tyk Dashboard to configure the Mock Response middleware for your Tyk Classic API by following these steps.

#### Step 1: Add an endpoint for the path and configure a list plugin
For the mock response to be enabled, the endpoint must also be in a list. We recommend adding the path to an allow list by [selecting]({{< ref "advanced-configuration/transform-traffic/endpoint-designer#allowlist" >}}) the **Whitelist** plugin.

> < TODO: add link to new allow_list middleware which should be here: product-stack/tyk-gateway/middleware/allow-list-tyk-classic#configuring-the-allow-list-in-the-api-designer >

Reviewer: Remember the TODO

#### Step 2: Add the mock response plugin
Now select the **Mock response** plugin.

> < TODO: add image >

Reviewer: Remember to add the image

#### Step 3: Configure the middleware

> Once you have selected the Mock response middleware for the endpoint, you can configure the HTTP status code, headers and body to be included in the response. Remember to click **ADD** to add each header to the response.

Reviewer: ```suggestion
Once you have selected the Mock response middleware for the endpoint, you can configure the HTTP status code, headers and body to be included in the response. Remember to click **ADD**, to add each header to the response.
```

> < TODO: add image >

Reviewer: Remember to add image

If you're using the newer Tyk OAS APIs, then check out the [Tyk OAS]({{< ref "product-stack/tyk-gateway/middleware/mock-response-tyk-oas" >}}) page.

## Configuring the middleware in the Tyk Classic API Definition

> To enable mock response, you must first add the endpoint to a list - one of [allow list]({{< ref "product-stack/tyk-gateway/middleware/allow-list-middleware" >}}), [block list]({{< ref "product-stack/tyk-gateway/middleware/block-list-middleware" >}}) or [ignore authentication]({{< ref "product-stack/tyk-gateway/middleware/ignore-middleware" >}}). This will add a new object to the `extended_paths` section of your API definition - `white_list`, `black_list` or `ignored`. The mock response can then be configured within the `method_actions` element within the new object.

Reviewer: Should _white_list_ and _black_list_ be replaced by allow list and block list respectively, or are these fields automatically generated from allow list and black list?

Author: This is behaviour of the product (which is what these instructions are documenting)
